File: path-sum-ii/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        '''
        Given a binary tree and a target sum, find paths root->leaf that equals the targetSum
        Returns a list of paths root->leaf as a list of lists
        '''
        answer = []
        if not root:
            return answer
		
        self.find_current_paths(root, targetSum, [], answer)
        return answer

    def find_current_paths(self, root: TreeNode, s: int, temp: list[int], answer: list[list[int]]):
        if not root:
            return
        
        temp.append(root.val)
        if root.val == s and not root.left and not root.right:
            answer.append(list(temp))
        else:
            self.find_current_paths(root.left, s-root.val, temp, answer)
            self.find_current_paths(root.right, s-root.val, temp, answer)
        del temp[-1]

File: path-sum-ii/path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) 
    {
        List<List<Integer>> answer = new ArrayList<>();
        if (root == null) 
            return answer;
        
        DFS(root, targetSum, answer, new ArrayList<>());
        
        return answer;
    }
    
    private void DFS(TreeNode node, int target, List<List<Integer>> answer, List<Integer> current)
    {
        // No node exists
        if (node == null) return;
        
        // Target value reached & is a leaf node
        else if (target - node.val == 0 && node.left == null && node.right == null)
        {
            current.add(node.val);
            answer.add(new ArrayList<>(current));
            current.remove(current.size()-1);
            return;
        }
        
        // Perform DFS and keep track of values seen
        current.add(node.val);
        DFS(node.left, target - node.val, answer, current);
        DFS(node.right, target - node.val, answer, current);
        current.remove(current.size()-1);
    }
}